File: services/workplaces/tavern/lib/tavern/dapr.ex
defmodule Tavern.Dapr do
  @moduledoc """
  The slice of the Dapr HTTP API this service needs, written by hand.

  There is no Elixir SDK for Dapr, which is the whole reason this module is
  interesting: it is the check on whether "a new building type is an hour of
  work in any language" survives contact with a runtime that ships SDKs for
  five languages and not this one.

  The answer is that it does, and the bill is small. Actor state is three HTTP
  calls (`GET .../state/<key>`, `POST .../state`, and invocation), the entity
  declaration is a JSON body at a fixed path, and the callbacks are two routes
  on a `Plug` router that already existed. `:httpc` from OTP serves all of it,
  so this adds no dependency — the same argument `Tavern.Connect` makes about
  not taking a gRPC library to speak a protocol that is really just POST.

  What the SDKs would have provided and this does not: typed actor proxies,
  reminders, and timers. Only reminders would be missed here, and Dapr 1.15
  needs a separate scheduler service for those anyway.
  """

  require Logger

  @actor_type "tavern"
  @state_key "shifts"

  def actor_type, do: @actor_type
  def state_key, do: @state_key

  @doc """
  Where this pod's sidecar is, or `nil` if there is none.

  `DAPR_HTTP_PORT` is injected by the sidecar itself, so its presence is the
  signal — there is no separate flag to get out of step with reality.
  """
  @spec sidecar() :: String.t() | nil
  def sidecar do
    case System.get_env("DAPR_HTTP_PORT") do
      nil -> nil
      "" -> nil
      port -> "http://localhost:#{port}"
    end
  end

  @doc "Reads one building's state. A miss is an empty building, not an error."
  @spec get_state(String.t(), non_neg_integer()) :: {:ok, map()} | {:error, term()}
  def get_state(base, building) do
    url = "#{base}/v1.0/actors/#{@actor_type}/#{building}/state/#{@state_key}"

    case request(:get, url, nil) do
      # 204 is "this actor has never saved anything", which is the same world
      # as "nobody is working here". Telling them apart would make the first
      # claim of a building's life fail.
      {:ok, 204, _body} -> {:ok, %{}}
      {:ok, 404, _body} -> {:ok, %{}}
      {:ok, 200, ""} -> {:ok, %{}}
      {:ok, 200, body} -> decode_state(body)
      {:ok, status, body} -> {:error, {:state_get, status, body}}
      {:error, reason} -> {:error, reason}
    end
  end

  @doc "Writes one building's state through the transactional endpoint."
  @spec put_state(String.t(), non_neg_integer(), map()) :: :ok | {:error, term()}
  def put_state(base, building, shifts) do
    url = "#{base}/v1.0/actors/#{@actor_type}/#{building}/state"

    body =
      Jason.encode!([
        %{
          "operation" => "upsert",
          "request" => %{"key" => @state_key, "value" => encode_state(shifts)}
        }
      ])

    case request(:post, url, body) do
      {:ok, status, _} when status < 300 -> :ok
      {:ok, status, reply} -> {:error, {:state_post, status, reply}}
      {:error, reason} -> {:error, reason}
    end
  end

  @doc """
  Invokes a method on a building's actor, going out through the sidecar.

  This is what earns the right to write: Dapr answers the state API with
  ERR_ACTOR_INSTANCE_MISSING unless the caller is executing inside an
  invocation it routed, so the trip out and back is not indirection for its own
  sake.
  """
  @spec invoke(String.t(), non_neg_integer(), String.t(), iodata()) ::
          {:ok, binary()} | {:error, term()}
  def invoke(base, building, method, body) do
    url = "#{base}/v1.0/actors/#{@actor_type}/#{building}/method/#{method}"

    case request(:post, url, body) do
      {:ok, status, reply} when status < 300 -> {:ok, reply}
      {:ok, status, reply} -> {:error, {:invoke, method, status, reply}}
      {:error, reason} -> {:error, reason}
    end
  end

  @doc """
  What the sidecar polls at startup to learn which entities this app hosts.

  Get it wrong and actors are never registered, which then fails as "actor
  instance is missing" a long way from the cause.
  """
  def config do
    %{
      entities: [@actor_type],
      # An hour, because a building is not idle in the sense Dapr means: it
      # exists whether or not anyone is drinking in it, and deactivating one
      # only costs a reload on the next call.
      actorIdleTimeout: "1h",
      actorScanInterval: "30s",
      drainOngoingCallTimeout: "30s",
      drainRebalancedActors: true
    }
  end

  # --- wire form -------------------------------------------------------------

  # Shift structs are the tavern's; the JSON shape is this module's. Tagging the
  # struct instead would let the store's format leak into the rules.
  defp encode_state(shifts) do
    Map.new(shifts, fn {pip, s} ->
      {to_string(pip),
       %{
         "started_tick" => s.started_tick,
         "last_work_tick" => s.last_work_tick,
         "last_work_ms" => s.last_work_ms
       }}
    end)
  end

  defp decode_state(body) do
    with {:ok, raw} <- Jason.decode(body) do
      {:ok,
       Map.new(raw, fn {pip, s} ->
         {String.to_integer(pip),
          %Tavern.Shifts.Shift{
            started_tick: s["started_tick"],
            last_work_tick: s["last_work_tick"],
            last_work_ms: s["last_work_ms"]
          }}
       end)}
    end
  rescue
    e -> {:error, {:state_decode, e}}
  end

  # --- transport -------------------------------------------------------------

  # :httpc rather than a client library. Adds nothing to mix.exs, and what is
  # being spoken here is POST with a JSON body — the same argument Tavern.Connect
  # makes for not taking a gRPC stack to serve unary Connect.
  defp request(method, url, body) do
    request =
      case method do
        :get -> {String.to_charlist(url), []}
        :post -> {String.to_charlist(url), [], ~c"application/json", body || ""}
      end

    case :httpc.request(method, request, [{:timeout, 5_000}], body_format: :binary) do
      {:ok, {{_version, status, _reason}, _headers, reply}} ->
        {:ok, status, reply}

      {:error, reason} ->
        {:error, reason}
    end
  end
end

defmodule Tavern.Store.Dapr do
  @moduledoc """
  A building's shifts, in the Dapr actor state store.

  Usable only from inside an actor invocation — see `Tavern.Dapr.invoke/4` for
  why. The Connect handler therefore does not call this; it forwards, and the
  actor endpoint on the other side is what lands here.

  No compare-and-set and no script. The `GenServer` store gets its atomicity
  from the process and the farm's Redis store gets it from Lua; here the actor
  runtime allows one invocation at a time per building, so read-modify-write is
  already indivisible.
  """

  @behaviour Tavern.Store

  alias Tavern.Dapr
  alias Tavern.Shifts
  alias Tavern.ShiftSet

  require Logger

  @impl true
  def claim(building, pip, tick) do
    case load(building) do
      {:error, reason} ->
        {:error, inspect(reason)}

      {:ok, shifts} ->
        claim_in(building, shifts, pip, tick)
    end
  end

  defp claim_in(building, shifts, pip, tick) do
    case ShiftSet.claim(shifts, pip, tick, now(), Shifts.rules()) do
      {:ok, updated, expired} ->
        log_expired(expired)
        save(building, updated)

      {:error, reason, _updated, expired} ->
        # Deliberately not written back. Offers are rejected constantly once a
        # building is full, and persisting the unchanged set on each one would
        # turn a busy tavern into a write storm.
        log_expired(expired)
        {:error, reason}
    end
  end

  @impl true
  def touch(building, pip, tick) do
    with {:ok, shifts} <- load(building),
         {:ok, elapsed, updated} <- ShiftSet.touch(shifts, pip, tick, now(), Shifts.rules()) do
      # Written even when elapsed is zero: Touch is what renews the lease, and
      # skipping the write would let a shift expire while it is being worked.
      case save(building, updated) do
        :ok -> {:ok, elapsed}
        error -> error
      end
    else
      :not_on_shift -> :not_on_shift
      {:error, reason} -> {:error, inspect(reason)}
    end
  end

  @impl true
  def release(building, pip) do
    with {:ok, shifts} <- load(building),
         {:ok, started, updated} <- ShiftSet.release(shifts, pip) do
      case save(building, updated) do
        :ok -> {:ok, started}
        error -> error
      end
    else
      :not_on_shift -> :not_on_shift
      {:error, reason} -> {:error, inspect(reason)}
    end
  end

  @impl true
  def count(building) do
    case load(building) do
      {:ok, shifts} ->
        map_size(shifts)

      {:error, reason} ->
        Logger.warning("could not count shifts", building: building, error: inspect(reason))
        0
    end
  end

  defp load(building), do: Dapr.get_state(base(), building)
  defp save(building, shifts), do: Dapr.put_state(base(), building, shifts)

  defp base do
    Dapr.sidecar() || raise "Tavern.Store.Dapr used without a sidecar"
  end

  # Wall clock, not monotonic — and the difference matters here in a way it does
  # not for the in-process store.
  #
  # `Tavern.Shifts` uses monotonic time, which is right for state that dies with
  # the process. This state outlives it: a lease written before a restart is
  # compared against a reading taken after one, and monotonic time has no
  # meaning across that boundary. It would leave every shift either immortal or
  # instantly expired, depending on which way the origin moved.
  defp now, do: System.system_time(:millisecond)

  defp log_expired(pips) do
    Enum.each(pips, fn pip -> Logger.info("shift lease expired", pip: pip) end)
  end
end

File: services/workplaces/farm/internal/farm/dapr.go
package farm

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"strconv"
	"strings"
	"time"

	"connectrpc.com/connect"
	"google.golang.org/protobuf/encoding/protojson"
	"google.golang.org/protobuf/proto"

	workplacev1 "github.com/teceer/pipsim/gen/go/pips/workplace/v1"
)

// ActorType is the Dapr entity name a farm building is registered under.
const ActorType = "farm"

// stateKey holds the whole shift set for one building. One key rather than one
// per pip: capacity is a property of the set, so counting has to see all of it
// at once, and the actor runtime already serialises access.
const stateKey = "shifts"

// --- state -------------------------------------------------------------------

// daprStore keeps one building's shifts in the Dapr actor state store.
//
// It is only usable from inside an actor invocation. Dapr answers the state API
// with ERR_ACTOR_INSTANCE_MISSING anywhere else, because activation is what
// makes this process authoritative for the id — which is precisely the property
// being bought, and why the Connect handlers forward instead of calling here.
//
// No Lua, and no compare-and-set. The Redis store needs an atomic
// reap-check-claim because several replicas share one key; here the actor
// runtime guarantees one invocation at a time per building, so a plain
// read-modify-write is already indivisible.
type daprStore struct {
	client *http.Client
	base   string
	id     uint64

	max    int
	lease  time.Duration
	maxGap uint64
	now    func() time.Time
}

// NewDaprStore builds a store backed by the actor state store behind the
// sidecar at base (e.g. "http://localhost:3500").
func NewDaprStore(base string, workplaceID uint64, max int, lease time.Duration, maxGap uint64) Store {
	return &daprStore{
		client: &http.Client{Timeout: 5 * time.Second},
		base:   strings.TrimSuffix(base, "/"),
		id:     workplaceID,
		max:    max,
		lease:  lease,
		maxGap: maxGap,
		now:    time.Now,
	}
}

// wireShift is the persisted form. shiftSet's fields are unexported on purpose —
// the rules own them — so the wire shape is written out here rather than by
// tagging the domain type and letting the store's format leak into it.
type wireShift struct {
	StartedTick  uint64    `json:"started_tick"`
	LastWorkTick uint64    `json:"last_work_tick"`
	LastWork     time.Time `json:"last_work"`
}

func (d *daprStore) stateURL(suffix string) string {
	return fmt.Sprintf("%s/v1.0/actors/%s/%d/state%s", d.base, ActorType, d.id, suffix)
}

func (d *daprStore) load(ctx context.Context) (*shiftSet, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, d.stateURL("/"+stateKey), nil)
	if err != nil {
		return nil, err
	}
	res, err := d.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer res.Body.Close()

	set := newShiftSet(d.max, d.lease, d.maxGap, d.now)

	// A miss is an empty building, not an error: an actor exists as soon as it
	// is addressed, so "never written" and "nobody working here" are the same
	// state and the caller must not be able to tell them apart.
	if res.StatusCode == http.StatusNoContent || res.StatusCode == http.StatusNotFound {
		return set, nil
	}
	body, _ := io.ReadAll(res.Body)
	if res.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("actor state get: %d %s", res.StatusCode, body)
	}
	if len(bytes.TrimSpace(body)) == 0 {
		return set, nil
	}

	var wire map[uint64]wireShift
	if err := json.Unmarshal(body, &wire); err != nil {
		return nil, fmt.Errorf("actor state decode: %w", err)
	}
	for pip, w := range wire {
		set.shifts[pip] = &shift{
			startedTick:  w.StartedTick,
			lastWorkTick: w.LastWorkTick,
			lastWork:     w.LastWork,
		}
	}
	return set, nil
}

func (d *daprStore) save(ctx context.Context, set *shiftSet) error {
	wire := make(map[uint64]wireShift, len(set.shifts))
	for pip, sh := range set.shifts {
		wire[pip] = wireShift{
			StartedTick:  sh.startedTick,
			LastWorkTick: sh.lastWorkTick,
			LastWork:     sh.lastWork,
		}
	}

	body, err := json.Marshal([]map[string]any{{
		"operation": "upsert",
		"request":   map[string]any{"key": stateKey, "value": wire},
	}})
	if err != nil {
		return err
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, d.stateURL(""), bytes.NewReader(body))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	res, err := d.client.Do(req)
	if err != nil {
		return err
	}
	defer res.Body.Close()
	if res.StatusCode >= 300 {
		msg, _ := io.ReadAll(res.Body)
		return fmt.Errorf("actor state post: %d %s", res.StatusCode, msg)
	}
	return nil
}

func (d *daprStore) Claim(ctx context.Context, pip, tick uint64) (bool, string, error) {
	set, err := d.load(ctx)
	if err != nil {
		return false, "", err
	}
	accepted, reason := set.claim(pip, tick)
	if !accepted {
		// Nothing changed except possibly a reap, and writing that back on every
		// rejected offer would turn a busy building into a write storm.
		return false, reason, nil
	}
	if err := d.save(ctx, set); err != nil {
		return false, "", err
	}
	return true, "", nil
}

func (d *daprStore) Touch(ctx context.Context, pip, tick uint64) (int32, bool, error) {
	set, err := d.load(ctx)
	if err != nil {
		return 0, false, err
	}
	elapsed, ok := set.touch(pip, tick)
	if !ok {
		return 0, false, nil
	}
	// Written even when elapsed is zero: Touch is what renews the lease, and
	// skipping the write would let a shift expire while it is being worked.
	if err := d.save(ctx, set); err != nil {
		return 0, false, err
	}
	return elapsed, true, nil
}

func (d *daprStore) Release(ctx context.Context, pip, _ uint64) (uint64, bool, error) {
	set, err := d.load(ctx)
	if err != nil {
		return 0, false, err
	}
	started, found := set.release(pip)
	if !found {
		return 0, false, nil
	}
	if err := d.save(ctx, set); err != nil {
		return 0, false, err
	}
	return started, true, nil
}

func (d *daprStore) Count(ctx context.Context) (int, error) {
	set, err := d.load(ctx)
	if err != nil {
		return 0, err
	}
	return set.count(), nil
}

// --- the two halves of a Dapr-hosted workplace -------------------------------

// ActorHost serves pips.workplace.v1 by handing every call to a building's
// actor, and serves the actor endpoints that answer on the other side.
//
// Both halves live in this process and talk through the sidecar. That loop
// looks wasteful and is not optional: state may only be touched inside an
// invocation the sidecar routed, so the trip out and back is what earns the
// right to write. What it buys is one invocation at a time per building —
// the atomicity store.go's Lua exists to provide — plus placement, so which
// replica owns which building stops being this service's problem.
//
// Costs roughly 1.6 ms per call, measured. Irrelevant at the economy's 1 Hz;
// it would not be if anything called a workplace per tick.
type ActorHost struct {
	inner  *Host
	client *http.Client
	base   string
}

func NewActorHost(inner *Host, base string) *ActorHost {
	return &ActorHost{
		inner:  inner,
		client: &http.Client{Timeout: 10 * time.Second},
		base:   strings.TrimSuffix(base, "/"),
	}
}

// call routes one RPC through the sidecar to the building's actor.
//
// The body is the protobuf request as JSON and the reply is the protobuf
// response as JSON, so the actor endpoint stays a bridge rather than a second
// hand-written encoding of the contract that could disagree with the first.
func (a *ActorHost) call(ctx context.Context, id uint64, method string, in, out proto.Message) error {
	body, err := protojson.Marshal(in)
	if err != nil {
		return err
	}

	url := fmt.Sprintf("%s/v1.0/actors/%s/%d/method/%s", a.base, ActorType, id, method)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")

	res, err := a.client.Do(req)
	if err != nil {
		return err
	}
	defer res.Body.Close()
	raw, _ := io.ReadAll(res.Body)
	if res.StatusCode >= 300 {
		return fmt.Errorf("actor %s/%d %s: %d %s", ActorType, id, method, res.StatusCode, raw)
	}
	if len(bytes.TrimSpace(raw)) == 0 {
		return nil
	}
	return protojson.Unmarshal(raw, out)
}

// resolveID fills in the building an RPC is for, applying the same
// single-building courtesy Host does so a Dapr-hosted farm behaves identically
// to a plain one.
func (a *ActorHost) resolveID(id uint64) (uint64, error) {
	b, err := a.inner.resolve(id)
	if err != nil {
		return 0, err
	}
	return b.id, nil
}

func (a *ActorHost) List(
	ctx context.Context,
	req *connect.Request[workplacev1.ListRequest],
) (*connect.Response[workplacev1.ListResponse], error) {
	// Enumeration is configuration, not state, so it is answered locally. Only
	// the occupancy inside each Describe needs an actor, and Describe below is
	// what fetches it.
	ids := a.inner.ids()
	out := make([]*workplacev1.DescribeResponse, 0, len(ids))
	for _, id := range ids {
		res, err := a.Describe(ctx, connect.NewRequest(&workplacev1.DescribeRequest{
			WorkplaceId: id,
		}))
		if err != nil {
			return nil, err
		}
		out = append(out, res.Msg)
	}
	_ = req
	return connect.NewResponse(&workplacev1.ListResponse{Workplaces: out}), nil
}

func (a *ActorHost) Describe(
	ctx context.Context,
	req *connect.Request[workplacev1.DescribeRequest],
) (*connect.Response[workplacev1.DescribeResponse], error) {
	id, err := a.resolveID(req.Msg.GetWorkplaceId())
	if err != nil {
		return nil, err
	}
	msg := req.Msg
	msg.WorkplaceId = id

	out := &workplacev1.DescribeResponse{}
	if err := a.call(ctx, id, "Describe", msg, out); err != nil {
		return nil, connect.NewError(connect.CodeUnavailable, err)
	}
	return connect.NewResponse(out), nil
}

func (a *ActorHost) CanEmploy(
	ctx context.Context,
	req *connect.Request[workplacev1.CanEmployRequest],
) (*connect.Response[workplacev1.CanEmployResponse], error) {
	id, err := a.resolveID(req.Msg.GetWorkplaceId())
	if err != nil {
		return nil, err
	}
	msg := req.Msg
	msg.WorkplaceId = id

	out := &workplacev1.CanEmployResponse{}
	if err := a.call(ctx, id, "CanEmploy", msg, out); err != nil {
		return nil, connect.NewError(connect.CodeUnavailable, err)
	}
	return connect.NewResponse(out), nil
}

func (a *ActorHost) StartShift(
	ctx context.Context,
	req *connect.Request[workplacev1.StartShiftRequest],
) (*connect.Response[workplacev1.StartShiftResponse], error) {
	id, err := a.resolveID(req.Msg.GetWorkplaceId())
	if err != nil {
		return nil, err
	}
	msg := req.Msg
	msg.WorkplaceId = id

	out := &workplacev1.StartShiftResponse{}
	if err := a.call(ctx, id, "StartShift", msg, out); err != nil {
		return nil, connect.NewError(connect.CodeUnavailable, err)
	}
	return connect.NewResponse(out), nil
}

func (a *ActorHost) Work(
	ctx context.Context,
	req *connect.Request[workplacev1.WorkRequest],
) (*connect.Response[workplacev1.WorkResponse], error) {
	id, err := a.resolveID(req.Msg.GetWorkplaceId())
	if err != nil {
		return nil, err
	}
	msg := req.Msg
	msg.WorkplaceId = id

	out := &workplacev1.WorkResponse{}
	if err := a.call(ctx, id, "Work", msg, out); err != nil {
		return nil, connect.NewError(connect.CodeUnavailable, err)
	}
	return connect.NewResponse(out), nil
}

func (a *ActorHost) EndShift(
	ctx context.Context,
	req *connect.Request[workplacev1.EndShiftRequest],
) (*connect.Response[workplacev1.EndShiftResponse], error) {
	id, err := a.resolveID(req.Msg.GetWorkplaceId())
	if err != nil {
		return nil, err
	}
	msg := req.Msg
	msg.WorkplaceId = id

	out := &workplacev1.EndShiftResponse{}
	if err := a.call(ctx, id, "EndShift", msg, out); err != nil {
		return nil, connect.NewError(connect.CodeUnavailable, err)
	}
	return connect.NewResponse(out), nil
}

func (a *ActorHost) Buy(
	ctx context.Context,
	req *connect.Request[workplacev1.BuyRequest],
) (*connect.Response[workplacev1.BuyResponse], error) {
	id, err := a.resolveID(req.Msg.GetWorkplaceId())
	if err != nil {
		return nil, err
	}
	msg := req.Msg
	msg.WorkplaceId = id

	out := &workplacev1.BuyResponse{}
	if err := a.call(ctx, id, "Buy", msg, out); err != nil {
		return nil, connect.NewError(connect.CodeUnavailable, err)
	}
	return connect.NewResponse(out), nil
}

// ConsiderOffer answers a queued offer. It goes through an actor like anything
// else that touches state — the offer arrives on a queue rather than an RPC,
// but the write it causes is the same write.
func (a *ActorHost) ConsiderOffer(ctx context.Context, pipID, tick uint64) (bool, string) {
	for _, id := range a.inner.ids() {
		out := &workplacev1.StartShiftResponse{}
		err := a.call(ctx, id, "StartShift", &workplacev1.StartShiftRequest{
			WorkplaceId: id, PipId: pipID, Tick: tick,
		}, out)
		if err != nil {
			slog.Warn("could not offer to a building", "workplace", id, "pip", pipID, "err", err)
			continue
		}
		if out.GetAccepted() {
			slog.Info("offer accepted", "workplace", id, "pip", pipID, "tick", tick)
			return true, ""
		}
	}
	return false, "no free positions"
}

// Workers counts through the actors, not through the buildings.
//
// Delegating to the inner Host would read the state store directly, and outside
// an invocation Dapr refuses that with ERR_ACTOR_INSTANCE_MISSING — so /healthz
// would report a failure on a service that is perfectly healthy. Everything
// that touches state goes the long way round; there is no shortcut for the
// health endpoint either.
func (a *ActorHost) Workers() int {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	total := 0
	for _, id := range a.inner.ids() {
		out := &workplacev1.DescribeResponse{}
		if err := a.call(ctx, id, "Describe",
			&workplacev1.DescribeRequest{WorkplaceId: id}, out); err != nil {
			slog.Warn("could not count workers", "workplace", id, "err", err)
			continue
		}
		total += int(out.GetCurrentWorkers())
	}
	return total
}

// --- what Dapr calls back into -----------------------------------------------

// Handler serves the endpoints the sidecar requires: the entity declaration it
// polls at startup, and the invocations it routes to activated actors.
//
// Mounted on the same mux as the Connect handler. They cannot collide: Connect
// owns /pips.workplace.v1.…, these own /dapr/ and /actors/.
func (h *Host) Handler() http.Handler {
	mux := http.NewServeMux()

	mux.HandleFunc("/dapr/config", func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(map[string]any{
			"entities": []string{ActorType},
			// An hour, because a building is not idle in the sense Dapr means:
			// it exists whether or not anyone is working in it, and
			// deactivating one only costs a reload on the next call.
			"actorIdleTimeout":  "1h",
			"actorScanInterval": "30s",
			// Long enough for a Work call to finish rather than being cut in
			// half during a rebalance.
			"drainOngoingCallTimeout": "30s",
			"drainRebalancedActors":   true,
		})
	})

	mux.HandleFunc("/actors/", h.serveActor)
	return mux
}

// serveActor dispatches one actor invocation to the building it names.
//
// This is where business logic legally runs: the sidecar routed the call, so
// the state API will accept writes for this id.
func (h *Host) serveActor(w http.ResponseWriter, r *http.Request) {
	// /actors/<type>/<id>[/method/<name>]
	parts := strings.Split(strings.Trim(r.URL.Path, "/"), "/")
	if len(parts) < 3 {
		http.Error(w, "bad actor path", http.StatusBadRequest)
		return
	}
	id, err := strconv.ParseUint(parts[2], 10, 64)
	if err != nil {
		http.Error(w, "bad actor id", http.StatusBadRequest)
		return
	}

	// Deactivation, and the activation probe Dapr sends with no method. Neither
	// needs anything done: a building keeps its state in the store, so there is
	// nothing in memory to build up or tear down.
	if r.Method == http.MethodDelete || len(parts) < 5 {
		w.WriteHeader(http.StatusOK)
		return
	}

	b, err := h.resolve(id)
	if err != nil {
		http.Error(w, err.Error(), http.StatusNotFound)
		return
	}

	body, _ := io.ReadAll(r.Body)
	reply, err := dispatch(r.Context(), b, parts[4], body)
	if err != nil {
		// 500 rather than a Connect code: the caller here is the sidecar, and
		// the adapter on the other side turns this into Unavailable.
		slog.Warn("actor invocation failed",
			"workplace", id, "method", parts[4], "err", err)
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	_, _ = w.Write(reply)
}

// dispatch runs one RPC against a building and returns its response as JSON.
func dispatch(ctx context.Context, b *Service, method string, body []byte) ([]byte, error) {
	unmarshal := func(m proto.Message) error {
		if len(bytes.TrimSpace(body)) == 0 {
			return nil
		}
		return protojson.Unmarshal(body, m)
	}

	switch method {
	case "Describe":
		in := &workplacev1.DescribeRequest{}
		if err := unmarshal(in); err != nil {
			return nil, err
		}
		res, err := b.Describe(ctx, connect.NewRequest(in))
		if err != nil {
			return nil, err
		}
		return protojson.Marshal(res.Msg)

	case "CanEmploy":
		in := &workplacev1.CanEmployRequest{}
		if err := unmarshal(in); err != nil {
			return nil, err
		}
		res, err := b.CanEmploy(ctx, connect.NewRequest(in))
		if err != nil {
			return nil, err
		}
		return protojson.Marshal(res.Msg)

	case "StartShift":
		in := &workplacev1.StartShiftRequest{}
		if err := unmarshal(in); err != nil {
			return nil, err
		}
		res, err := b.StartShift(ctx, connect.NewRequest(in))
		if err != nil {
			return nil, err
		}
		return protojson.Marshal(res.Msg)

	case "Work":
		in := &workplacev1.WorkRequest{}
		if err := unmarshal(in); err != nil {
			return nil, err
		}
		res, err := b.Work(ctx, connect.NewRequest(in))
		if err != nil {
			return nil, err
		}
		return protojson.Marshal(res.Msg)

	case "EndShift":
		in := &workplacev1.EndShiftRequest{}
		if err := unmarshal(in); err != nil {
			return nil, err
		}
		res, err := b.EndShift(ctx, connect.NewRequest(in))
		if err != nil {
			return nil, err
		}
		return protojson.Marshal(res.Msg)

	case "Buy":
		in := &workplacev1.BuyRequest{}
		if err := unmarshal(in); err != nil {
			return nil, err
		}
		res, err := b.Buy(ctx, connect.NewRequest(in))
		if err != nil {
			return nil, err
		}
		return protojson.Marshal(res.Msg)

	default:
		return nil, fmt.Errorf("unknown actor method %q", method)
	}
}
